cellar_rebalancer/src/cellars/uniswapv3.rs

```diff
@@ -158,11 +166,44 @@ impl<T: 'static + Middleware> UniswapV3CellarState<T> {
     }
 }
 
-// Struct for UniswapV3CellarTickInfo
-#[derive(Clone, Debug)]
-pub struct UniswapV3CellarTickInfo {
-    pub(crate) token_id: U256,
-    pub(crate) tick_upper: i32,
-    pub(crate) tick_lower: i32,
-    pub(crate) weight: u32,
+pub struct UniswapV3CellarAllocator;
+
+#[async_trait]
+impl server::UniswapV3CellarAllocator for UniswapV3CellarAllocator {
+    async fn rebalance(
+        &self,
+        request: tonic::Request<RebalanceRequest>,
+    ) -> Result<tonic::Response<RebalanceResponse>, tonic::Status> {
+        let request = request.get_ref();
+        let tick_ranges: Vec<proto::TickRange> = request
+            .data
+            .clone()
+            .into_iter()
+            .map(|d| proto::TickRange {
+                upper: d.upper_price,
+                lower: d.lower_price,
+                weight: d.weight,
+            })
+            .collect();
+        let pair_id = APP
+            .config()
+            .cellars
+            // sketchy
+            .get(0)
+            .unwrap()
+            .pair_id;
+        let eth_gas_price = match cellars::get_gas_price().await {
+            Ok(gp) => gp,
+            Err(err) => {
+                error!("Failed to get cellar gas price: {:?}", err);
+                // what's the right way to handle this?
```

Comment:
Perhaps this belongs in `decide_rebalance` rather than the request handler. Additionally, you could add some fallbacks to Etherscan if you wanted to: `ethers` can look at price history in the chain to estimate gas price. Another alternative would be to have a config option for a default gas price.

Reply:
Whoops, I just realized, cosmos mode needs cosmos gas prices and not Ethereum gas prices!